`JNIEnv::new_local_ref` has seemingly-unnecessary lifetime restriction

The JNI `NewLocalRef` function can be used to create a local reference from a global reference. [The spec specifically allows it](https://docs.oracle.com/en/java/javase/11/docs/specs/jni/functions.html#newlocalref), and I'm guessing that's its main purpose. However, the wrapper method `JNIEnv::new_local_ref` currently can't be used that way because it requires the provided `JObject` to have the same lifetime as the `JNIEnv` itself (which won't be the case for a `JObject` belonging to a `GlobalRef`). This restriction does not seem necessary and should probably be removed (that is, the type of `obj` should be changed from `JObject<'a>` to `JObject<'_>`).

Also, `JNIEnv::new_local_ref` has a type parameter `T` that doesn't seem to be used at all.

No. I get a compile error because of the incorrect lifetime.

What I have is a cache of generated Java objects. Here's a generic version:

```rust
use jni::{
	errors::Result as JResult,
	JNIEnv,
	objects::{GlobalRef, JObject},
};
use std::{
	collections::{hash_map, HashMap},
	hash::Hash,
};

struct JCache<T> {
	refs: HashMap<T, GlobalRef>,
}

impl<T: Eq + Hash> JCache<T> {
	fn get<'e>(
		&mut self,
		env: JNIEnv<'e>,
		key: T,
		compute_jobject: impl FnOnce(JNIEnv<'e>, &T) -> JResult<JObject<'e>>,
	) -> JResult<JObject<'e>> {
		Ok(match self.refs.entry(key) {
			hash_map::Entry::Occupied(entry) => {
				let o: &GlobalRef = entry.get();
				let o = o.as_obj();
				let o: JObject<'e> = env.new_local_ref::<()>(o)?;
				o
			}
			hash_map::Entry::Vacant(entry) => {
				let o = compute_jobject(env, entry.key())?;
				entry.insert(env.new_global_ref(o)?);
				o
			}
		})
	}
}
```

This currently fails to compile:

```
error[E0623]: lifetime mismatch
  --> src/lib.rs:26:26
   |
17 |         &mut self,
   |         --------- these two types are declared with different lifetimes...
18 |         env: JNIEnv<'e>,
   |              ----------
...
26 |                 let o: JObject<'e> = env.new_local_ref::<()>(o)?;
   |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...but data from `self` flows into `env` here
```

If, however, the `new_local_ref` parameter's type is changed to `JObject<'_>`, then it compiles successfully.

My current workaround is to escape from lifetime constraints by turning the `JObject` into a raw pointer and then re-wrapping it with the correct lifetime:

```rust
let o = o.as_obj();
let o = JObject::<'e>::from(o.into_inner()); // workaround
let o: JObject<'e> = env.new_local_ref::<()>(o)?;
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -24,6 +24,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - The `release_string_utf_chars` function has been marked as unsafe. (#334)
 - Breaking change: Mark `JNIEnv::new_direct_byte_buffer` as `unsafe` (#320)
 - The lifetime of `AutoArray` is no longer tied to the lifetime of a particular `JNIEnv` reference. (#302)
+- Relaxed lifetime restrictions on `JNIEnv::new_local_ref`. Now it can be used to create a local
+  reference from a global reference. (#301 / #319)
 
 ## [0.19.0] — 2021-01-24
 

--- a/src/wrapper/jnienv.rs
+++ b/src/wrapper/jnienv.rs
@@ -373,11 +373,98 @@ impl<'a> JNIEnv<'a> {
         Ok(global)
     }
 
-    /// Create a new local ref to an object.
+    /// Create a new local reference to an object.
     ///
-    /// This is useful for creating a local reference from a [global reference][GlobalRef] or
-    /// creating a local reference in a different [local reference frame][JNIEnv::with_local_frame]
-    /// than the original.
+    /// Specifically, this calls the JNI function [`NewLocalRef`], which creates a reference in the
+    /// current local reference frame, regardless of whether the original reference belongs to the
+    /// same local reference frame, a different one, or is a [global reference][GlobalRef]. In Rust
+    /// terms, this method accepts a JNI reference with any valid lifetime and produces a clone of
+    /// that reference with the lifetime of this `JNIEnv`. The returned reference can outlive the
+    /// original.
+    ///
+    /// This method is useful when you have a strong global reference and you can't prevent it from
+    /// being dropped before you're finished with it. In that case, you can use this method to
+    /// create a new local reference that's guaranteed to remain valid for the duration of the
+    /// current local reference frame, regardless of what later happens to the original global
+    /// reference.
+    ///
+    /// # Lifetimes
+    ///
+    /// `'a` is the lifetime of this `JNIEnv`. This method creates a new local reference with
+    /// lifetime `'a`.
+    ///
+    /// `'b` is the lifetime of the original reference. It can be any valid lifetime, even one that
+    /// `'a` outlives or vice versa.
+    ///
+    /// Think of `'a` as meaning `'new` and `'b` as meaning `'original`. (It is unfortunately not
+    /// possible to actually give these names to the two lifetimes because `'a` is a parameter to
+    /// the `JNIEnv` type, not a parameter to this method.)
+    ///
+    /// # Example
+    ///
+    /// In the following example, the `ExampleError::extract_throwable` method uses
+    /// `JNIEnv::new_local_ref` to create a new local reference that outlives the original global
+    /// reference:
+    ///
+    /// ```no_run
+    /// # use jni::{JNIEnv, objects::*};
+    /// # use std::fmt::Display;
+    /// #
+    /// # type SomeOtherErrorType = Box<dyn Display>;
+    /// #
+    /// /// An error that may be caused by either a Java exception or something going wrong in Rust
+    /// /// code.
+    /// enum ExampleError {
+    ///     /// This variant represents a Java exception.
+    ///     ///
+    ///     /// The enclosed `GlobalRef` points to a Java object of class `java.lang.Throwable`
+    ///     /// (or one of its many subclasses).
+    ///     Exception(GlobalRef),
+    ///
+    ///     /// This variant represents an error in Rust code, not a Java exception.
+    ///     Other(SomeOtherErrorType),
+    /// }
+    ///
+    /// impl ExampleError {
+    ///     /// Consumes this `ExampleError` and produces a `JThrowable`, suitable for throwing
+    ///     /// back to Java code.
+    ///     ///
+    ///     /// If this is an `ExampleError::Exception`, then this extracts the enclosed Java
+    ///     /// exception object. Otherwise, a new exception object is created to represent this
+    ///     /// error.
+    ///     fn extract_throwable(self, env: JNIEnv) -> jni::errors::Result<JThrowable> {
+    ///         let throwable: JObject = match self {
+    ///             ExampleError::Exception(exception) => {
+    ///                 // The error was caused by a Java exception.
+    ///
+    ///                 // Here, `exception` is a `GlobalRef` pointing to a Java `Throwable`. It
+    ///                 // will be dropped at the end of this `match` arm. We'll use
+    ///                 // `new_local_ref` to create a local reference that will outlive the
+    ///                 // `GlobalRef`.
+    ///
+    ///                 env.new_local_ref(exception.as_obj())?
+    ///             }
+    ///
+    ///             ExampleError::Other(error) => {
+    ///                 // The error was caused by something that happened in Rust code. Create a
+    ///                 // new `java.lang.Error` to represent it.
+    ///
+    ///                 env.new_object(
+    ///                     "java/lang/Error",
+    ///                     "(Ljava/lang/String;)V",
+    ///                     &[
+    ///                         env.new_string(error.to_string())?.into(),
+    ///                     ],
+    ///                 )?
+    ///             }
+    ///         };
+    ///
+    ///         Ok(JThrowable::from(throwable))
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// [`NewLocalRef`]: https://docs.oracle.com/en/java/javase/11/docs/specs/jni/functions.html#newlocalref
     pub fn new_local_ref<'b, O>(&self, obj: O) -> Result<JObject<'a>>
     where
         O: Into<JObject<'b>>,

--- a/src/wrapper/jnienv.rs
+++ b/src/wrapper/jnienv.rs
@@ -375,11 +375,15 @@ impl<'a> JNIEnv<'a> {
 
     /// Create a new local ref to an object.
     ///
-    /// Note that the object passed to this is *already* a local ref. This
-    /// creates yet another reference to it, which is most likely not what you
-    /// want.
-    pub fn new_local_ref<T>(&self, obj: JObject<'a>) -> Result<JObject<'a>> {
-        let local: JObject = jni_unchecked!(self.internal, NewLocalRef, obj.into_inner()).into();
+    /// This is useful for creating a local reference from a [global reference][GlobalRef] or
+    /// creating a local reference in a different [local reference frame][JNIEnv::with_local_frame]
+    /// than the original.
+    pub fn new_local_ref<'b, O>(&self, obj: O) -> Result<JObject<'a>>
+    where
+        O: Into<JObject<'b>>,
+    {
+        let local: JObject =
+            jni_unchecked!(self.internal, NewLocalRef, obj.into().into_inner()).into();
         Ok(local)
     }
 

